hapi-fhir-jpaserver-hfql/src/test/java/ca/uhn/fhir/jpa/fql/parser/HfqlStatementParserTest.java

```diff
@@ -0,0 +1,500 @@
+package ca.uhn.fhir.jpa.fql.parser;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.parser.DataFormatException;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@TestMethodOrder(MethodOrderer.MethodName.class)
+@SuppressWarnings("SqlDialectInspection")
+public class HfqlStatementParserTest {
+
+	private static final FhirContext ourCtx = FhirContext.forR4Cached();
+
+	@Test
+	public void testCountAndGroup() {
+		String input = """
+   		select Count(*), name.given, name.family
+   		from Patient
+   		group by name.given, name.family
+			""";
+
+		HfqlStatement statement = parse(input);
+		assertEquals("Patient", statement.getFromResourceName());
+		assertEquals(3, statement.getSelectClauses().size());
+		assertEquals("*", statement.getSelectClauses().get(0).getClause());
+		assertEquals(HfqlStatement.SelectClauseOperator.COUNT, statement.getSelectClauses().get(0).getOperator());
+		assertEquals("Count(*)", statement.getSelectClauses().get(0).getAlias());
+		assertEquals("name.given", statement.getSelectClauses().get(1).getClause());
+		assertEquals(HfqlStatement.SelectClauseOperator.SELECT, statement.getSelectClauses().get(1).getOperator());
+		assertEquals("name.family", statement.getSelectClauses().get(2).getClause());
+		assertEquals(HfqlStatement.SelectClauseOperator.SELECT, statement.getSelectClauses().get(2).getOperator());
+		assertEquals(2, statement.getGroupByClauses().size());
+		assertThat(statement.getGroupByClauses(), contains("name.given", "name.family"));
+
+	}
+
+
+	@Test
+	public void testFromSelect() {
+		String input = """
+					from Patient
+					select
+					   name.given[0],
+					   name.family
+			""";
+
+		HfqlStatement statement = parse(input);
+		assertEquals("Patient", statement.getFromResourceName());
+		assertEquals(2, statement.getSelectClauses().size());
+		assertEquals("name.given[0]", statement.getSelectClauses().get(0).getClause());
+		assertEquals("name.given[0]", statement.getSelectClauses().get(0).getAlias());
+		assertEquals(HfqlStatement.SelectClauseOperator.SELECT, statement.getSelectClauses().get(0).getOperator());
+		assertEquals("name.family", statement.getSelectClauses().get(1).getClause());
+		assertEquals("name.family", statement.getSelectClauses().get(1).getAlias());
+	}
+
+	@Test
+	public void testSelect_ValueWithPrefix() {
+		String input = """
+					select id
+					from Observation
+					where value-quantity = 'lt500'
```

Comment:
suggestion: it is confusing to bounce between fhirpath and SP-names, and complicates the grammar.  How about a table-valued function:
```suggestion
					where _id in (search_match('Observation', 'value-quantity, 'lt500')
```
or a boolean predicate 
```suggestion
					where search_match(_id, 'Observation', 'value-quantity, 'lt500')
```
Or even fall back to url-form encoded
```
  where _id in (search_match('Observation?quantity=lt500')
```
It would be good to figure something out that works for joins easily.

Reply:
If we allow "row" aliases, and row parameters to our search "functions"/operators, then join could look like:
```
select o._id as obs_id, p._id as pat_id from Observation as o
JOIN Patient as p on search_reference_match(o, 'subject', p)
```

and I would write the WHERE clause as:
```
where search_match(o, 'quantity', 'lt500')
```

Reply:
This has been implemented as suggested.